File: java-accessibility-checker/src/main/java/com/ibm/able/equalaccess/report/ReporterManager.java
/******************************************************************************
    Copyright:: 2024- IBM, Inc

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
 *****************************************************************************/
package com.ibm.able.equalaccess.report;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.google.gson.Gson;
import com.ibm.able.equalaccess.abs.IAbstractAPI;
import com.ibm.able.equalaccess.config.ACConfigManager;
import com.ibm.able.equalaccess.config.ConfigInternal;
import com.ibm.able.equalaccess.engine.ACEReport;
import com.ibm.able.equalaccess.engine.ACError;
import com.ibm.able.equalaccess.engine.ACReport;
import com.ibm.able.equalaccess.engine.Guideline;
import com.ibm.able.equalaccess.engine.eRuleConfidence;
import com.ibm.able.equalaccess.engine.eRuleLevel;
import com.ibm.able.equalaccess.engine.ACReport.Result;
import com.ibm.able.equalaccess.enginecontext.EngineContextManager;
import com.ibm.able.equalaccess.enginecontext.IEngineContext;

public class ReporterManager {
    private static Gson gson = new Gson();
    private static ReporterManager singleton = null;
    public static ReporterManager get() {
        if (singleton == null) throw new ACError("ReporterManager not intialized");
        return singleton;
    }
    public static ReporterManager initialize(ConfigInternal config, IAbstractAPI absAPI, Guideline[] rulesets) {
        return singleton = new ReporterManager(config, absAPI, rulesets);
    }

    private ConfigInternal config;
    private IAbstractAPI absAPI;
    private Guideline[] rulesets;
    private List<IReporter> reporters = new ArrayList<>();
    private List<CompressedReport> reports = new ArrayList<>();
    private IReporter returnReporter = new ACReporterJSON();

    private ReporterManager(ConfigInternal config, IAbstractAPI absAPI, Guideline[] rulesets) {
        this.config = config;
        this.absAPI = absAPI;
        this.rulesets = rulesets;
        if (config.perfMetrics) {
            reporters.add(new ACReporterMetrics(config.toolName, config.policies));
        }

        if (!Arrays.asList(config.outputFormat).contains("disable")) {
            if (Arrays.asList(config.outputFormat).contains("json")) {
                reporters.add(new ACReporterJSON());
            }
            if (Arrays.asList(config.outputFormat).contains("html")) {
                // TODO:
                // reporters.push(new ACReporterHTML())
            }
            if (Arrays.asList(config.outputFormat).contains("csv")) {
                reporters.add(new ACReporterCSV(absAPI));
            }
            if (Arrays.asList(config.outputFormat).contains("xlsx")) {
                // TODO:
                // reporters.add(new ACReporterXLSX());
            }
        }
    }
    private Set<String> usedLabels = new HashSet<String>();

    public ACReport addEngineReport(String scanProfile, long startScan, String url, String pageTitle, String label, ACEReport engineReport) {
        verifyLabel(label);
        usedLabels.add(label);
        ACReport filteredReport = filterReport(engineReport, label);
        filteredReport.summary.startScan = startScan;
        filteredReport.summary.URL = url;
        filteredReport.label = label;
        ReporterStored storedReport = new ReporterStored(pageTitle, scanProfile, filteredReport);
        for (Result issue: filteredReport.results) {
            issue.help = getHelpUrl(issue);
        }        
        CompressedReport compressedReport = new CompressedReport(storedReport);
        if (reporters.size() > 0) {
            reports.add(compressedReport);
            for (IReporter reporter: reporters) {
                ReporterFile reportInfo = reporter.generateReport(config, rulesets, storedReport);
                if (reportInfo != null) {
                    try {
                        absAPI.writeFile(reportInfo.path, reportInfo.contents);
                    } catch (IOException e) {
                        System.err.println(e);
                        e.printStackTrace();
                    }
                }
            }
        }
        ReporterFile retVal = returnReporter.generateReport(config, rulesets, storedReport);
        if (retVal != null) return gson.fromJson(retVal.contents.toString(), ACReport.class);
        return null;
    }

    private void verifyLabel(String label) {
        // In the case that the label is null or undefined, throw an error
        if (label == null) {
            throw new ACError("labelNotProvided: Label must be provided when calling aChecker.getCompliance.");
        }

        // Check to make sure that the label that is provided is unique with all the other ones
        // that we have gone through.
        boolean labelUnique = isLabelUnique(label);

        // In the case that the label is not unique
        if (!labelUnique) {
            throw new ACError("Label provided to aChecker.getCompliance ("+label+") should be unique across all testcases in a single accessibility-checker session.");
        }
    }

    private boolean isLabelUnique(String label) {
        return !usedLabels.contains(label);
    }

    private ACReport filterReport(ACEReport engineResult, String scanLabel) {
        Map<String, Map<String, Set<String>>> ignoreLookup = new HashMap<String, Map<String, Set<String>>> ();

        ACReport baselineReport = absAPI.loadBaseline(scanLabel);
        if (baselineReport != null) {
            for (Result issue : baselineReport.results) {
                // ignoreLookup[issue.path.dom] = ignoreLookup[issue.path.dom] || {}
                Map<String, Set<String>> a = ignoreLookup.get(issue.path.get("dom"));
                if (a == null) {
                    ignoreLookup.put(issue.path.get("dom"), new HashMap<String, Set<String>>());
                    a = ignoreLookup.get(issue.path.get("dom"));
                }
                // ignoreLookup[issue.path.dom][issue.ruleId] = ignoreLookup[issue.path.dom][issue.ruleId] || {}
                Set<String> b = a.get(issue.ruleId);
                if (b == null) {
                    b = new HashSet<String>();
                }
                // ignoreLookup[issue.path.dom][issue.ruleId][issue.reasonId] = true;
                b.add(issue.reasonId);
            }
        }
        ACReport retVal = new ACReport(config, engineResult, scanLabel);

        // Set the config level and filter the results. Make note of which NLS keys we need
        for (ACReport.Result pageResult: retVal.results) {
            eRuleLevel reportLevel = ReporterManager.valueToLevel(pageResult.value);
            boolean ignored = false;
            if (!eRuleConfidence.PASS.toString().equals(pageResult.value[1]) 
                && ignoreLookup.containsKey(pageResult.path.get("dom")) 
                && ignoreLookup.get(pageResult.path.get("dom")).containsKey(pageResult.ruleId)
                && ignoreLookup.get(pageResult.path.get("dom")).get(pageResult.ruleId).contains(pageResult.reasonId))
            {
                ignored = true;
            }
            pageResult.ignored = ignored;
            pageResult.level = reportLevel;
        }

        retVal.addCounts(engineResult.summary.counts);
        retVal.filter(config.reportLevels);

        return retVal;
    }

    public static eRuleLevel valueToLevel(String[] reportValue) {
        eRuleLevel reportLevel = eRuleLevel.undefined;
        if ("PASS".equals(reportValue[1])) {
            reportLevel = eRuleLevel.pass;
        } else if (("VIOLATION".equals(reportValue[0]) || "RECOMMENDATION".equals(reportValue[0])) && "MANUAL".equals(reportValue[1])) {
            reportLevel = eRuleLevel.manual;
        } else if ("VIOLATION".equals(reportValue[0])) {
            if ("FAIL".equals(reportValue[1])) {
                reportLevel = eRuleLevel.violation;
            } else if ("POTENTIAL".equals(reportValue[1])) {
                reportLevel = eRuleLevel.potentialviolation;
            }
        } else if ("RECOMMENDATION".equals(reportValue[0])) {
            if ("FAIL".equals(reportValue[1])) {
                reportLevel = eRuleLevel.recommendation;
            } else if ("POTENTIAL".equals(reportValue[1])) {
                reportLevel = eRuleLevel.potentialrecommendation;
            }
        }
        return reportLevel;
    }

    private static final IEngineContext engine = EngineContextManager.getEngineContext(null);
    private String getHelpUrl(ACReport.Result issue) {
        if (issue.help != null && issue.help.length() > 0) return issue.help;
        ConfigInternal config = ACConfigManager.getConfigUnsupported();
        String helpUrl = engine.getHelp(issue.ruleId, issue.reasonId, config.ruleArchivePath == null ? config.ruleArchive : config.ruleArchivePath.substring(config.ruleArchivePath.lastIndexOf("/")+1));
        return helpUrl+"#"+engine.encodeURIComponent(issue.toHelpData());
    }

    public void generateSummaries() {
        long endReport = new Date().getTime();
        // If no scans, don't generate summaries
        if (reports.isEmpty()) return;
        for (IReporter reporter: reporters) {
            ReporterFile summaryInfo = reporter.generateSummary(config, rulesets, endReport, reports);
            if (summaryInfo != null) {
                try {
                    absAPI.writeFile(summaryInfo.path, summaryInfo.contents);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        reports.clear();
    }
}

File: karma-accessibility-checker/src/lib/ReporterManager.js
/******************************************************************************
     Copyright:: 2023- IBM, Inc

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*****************************************************************************/

let eRuleLevel = {
    violation: "violation",
    potentialviolation: "potentialviolation",
    recommendation: "recommendation",
    potentialrecommendation: "potentialrecommendation",
    manual: "manual",
    pass: "pass",
    ignored: "ignored"
}

let eRuleConfidence = {
    PASS: "PASS",
    FAIL: "FAIL",
    POTENTIAL: "POTENTIAL",
    MANUAL: "MANUAL"
}


/**
 *
 */
class ReporterManager {
    static usedLabels = {}
    static config;
    static checker;

    /**
     * This function is responsible for printing the scan results to string with the intent of that going to the console.
     *
     * @param {Object} results - Provide the results from the scan.
     *
     * @return {String} resultsString - String representation of the results/violations.
     *
     * PUBLIC API
     *
     * @memberOf this
     */
    static stringifyResults(reportP) {
        if (!reportP.results) {
            return `ERROR: ${JSON.stringify(reportP)}`;
        }
        const report = reportP;
        // console.log(report);
        // Variable Decleration
        let resultsString = `Scan: ${report.label}\n`;

        // Loop over the reports and build the string version of the the issues within each report
        report.results && report.results.forEach(function (issue) {
            if (ReporterManager.config.reportLevels.includes(issue.level)) {
                // Build string of the issues with only level, messageCode, xpath and snippet.
                resultsString += "- Message: " + issue.message +
                    "\n  Level: " + issue.level +
                    "\n  XPath: " + issue.path.dom +
                    "\n  Snippet: " + issue.snippet +
                    "\n  Help: " + issue.help +
                    "\n";
            }
        });

        return resultsString;
    };

    static getHelpUrl(issue) {
        if (issue.help) return issue.help;
        if (!ReporterManager.checker) {
            ReporterManager.checker = new ace.Checker();
        }
        let config = ReporterManager.config;
        let helpUrl = ReporterManager.checker.engine.getHelp(issue.ruleId, issue.reasonId, !config.ruleArchivePath ? config.ruleArchive : config.ruleArchivePath.substring(config.ruleArchivePath.lastIndexOf("/")+1));
        let minIssue = {
            message: issue.message,
            snippet: issue.snippet,
            value: issue.value,
            reasonId: issue.reasonId,
            ruleId: issue.ruleId,
            msgArgs: issue.messageArgs
        };
        return `${helpUrl}#${encodeURIComponent(JSON.stringify(minIssue))}`
    }

    static filterReport(engineResult, scanLabel) {
        ReporterManager.usedLabels[scanLabel] = true;
        let ignoreLookup = {}
        let baselineReport = aChecker.getBaseline(scanLabel);
        if (baselineReport) {
            for (const issue of baselineReport.results) {
                ignoreLookup[issue.path.dom] = ignoreLookup[issue.path.dom] || {}
                ignoreLookup[issue.path.dom][issue.ruleId] = ignoreLookup[issue.path.dom][issue.ruleId] || {}
                ignoreLookup[issue.path.dom][issue.ruleId][issue.reasonId] = true;
            }
        }
        let retVal = JSON.parse(JSON.stringify(engineResult));

        // Set the config level and filter the results. Make note of which NLS keys we need
        let keepNlsKeys = {}
        retVal.results = retVal.results.map(pageResult => {
            // Fetch the level from the results
            let reportValue = pageResult.value;
            let reportLevel;
            if (reportValue[1] === "PASS") {
                reportLevel = eRuleLevel.pass;
            } else if ((reportValue[0] === "VIOLATION" || reportValue[0] === "RECOMMENDATION") && reportValue[1] === "MANUAL") {
                reportLevel = eRuleLevel.manual;
            } else if (reportValue[0] === "VIOLATION") {
                if (reportValue[1] === "FAIL") {
                    reportLevel = eRuleLevel.violation;
                } else if (reportValue[1] === "POTENTIAL") {
                    reportLevel = eRuleLevel.potentialviolation;
                }
            } else if (reportValue[0] === "RECOMMENDATION") {
                if (reportValue[1] === "FAIL") {
                    reportLevel = eRuleLevel.recommendation;
                } else if (reportValue[1] === "POTENTIAL") {
                    reportLevel = eRuleLevel.potentialrecommendation;
                }
            }
            let ignored = false;
            if (pageResult.value[1] !== eRuleConfidence.PASS && pageResult.path.dom in ignoreLookup && pageResult.ruleId in ignoreLookup[pageResult.path.dom] && ignoreLookup[pageResult.path.dom][pageResult.ruleId][pageResult.reasonId]) {
                ignored = true;
            }
            return {
                ...pageResult,
                ignored,
                level: reportLevel
            }
        });

        retVal.summary = {};
        retVal.summary.counts = ReporterManager.getCounts(retVal);

        retVal.results = retVal.results.filter(pageResult => {
            if (ReporterManager.config.reportLevels.includes(pageResult.level)) {
                keepNlsKeys[pageResult.ruleId] = keepNlsKeys[pageResult.ruleId] || {
                    "0": true
                }
                keepNlsKeys[pageResult.ruleId][pageResult.reasonId] = true;
                if (pageResult.message.length > 32000) {
                    pageResult.message = pageResult.message.substring(0, 32000 - 3) + "...";
                }
                if (pageResult.snippet.length > 32000) {
                    pageResult.snippet = pageResult.snippet.substring(0, 32000 - 3) + "...";
                }
                return true;
            } else {
                return false;
            }
        });

        if (retVal.nls) {
            for (const ruleId in retVal.nls) {
                if (!(ruleId in keepNlsKeys)) {
                    delete retVal.nls[ruleId];
                } else {
                    for (const reasonId in retVal.nls[ruleId]) {
                        if (!(reasonId in keepNlsKeys[ruleId])) {
                            delete retVal.nls[ruleId][reasonId];
                        }
                    }
                }
            }
        }

        return retVal;
    }

    static getCounts(engineReport) {
        let counts = {
            violation: 0,
            potentialviolation: 0,
            recommendation: 0,
            potentialrecommendation: 0,
            manual: 0,
            pass: 0,
            ignored: 0,
            elements: 0,
            elementsViolation: 0,
            elementsViolationReview: 0
        }
        let elementSet = new Set();
        let elementViolationSet = new Set();
        let elementViolationReviewSet = new Set();
        for (const issue of engineReport.results) {
            elementSet.add(issue.path.dom);
            if (issue.ignored) {
                ++counts.ignored;
            } else {
                ++counts[issue.level.toString()];
                if (issue.level === eRuleLevel.violation) {
                    elementViolationSet.add(issue.path.dom);
                    elementViolationReviewSet.add(issue.path.dom);
                } else if (issue.level === eRuleLevel.potentialviolation || issue.level === eRuleLevel.manual) {
                    elementViolationReviewSet.add(issue.path.dom);
                }
            }
        }
        counts.elements = elementSet.size;
        counts.elementsViolation = elementViolationSet.size;
        counts.elementsViolationReview = elementViolationReviewSet.size
        return counts;
    }


    static isLabelUnique(label) {
        return !(label in ReporterManager.usedLabels);
    }

    static generateReport(config, rulesets, storedReport) {
        let outReport = JSON.parse(JSON.stringify(storedReport.engineReport));
        outReport.summary = ReporterManager.generateReportSummary(config, rulesets, storedReport);
        delete outReport.totalTime;
        outReport.scanID = config.scanID;
        outReport.toolID = config.toolID;
        outReport.label = storedReport.label;

        return {
            reportPath: `${storedReport.label}.json`,
            report: outReport
        };
    }

    static generateReportSummary(config, rulesets, storedReport) {
        let { engineReport, startScan, url } = storedReport;
        
        return {
            counts: engineReport.summary.counts,
            scanTime: engineReport.totalTime,
            ruleArchive: config.ruleArchiveLabel,
            policies: config.policies,
            reportLevels: config.reportLevels,
            startScan: startScan,
            URL: url
        }
    }
};
